### PR TITLE
feat(dashboards-v2): deterministic panel capability guard (type × query-type × signal)

### PR DESCRIPTION
**Pull request by ahrefabhi**
## Pull Request

> **Stacked on top of `fix/dashboards-v2-panel-editor`** — review/merge that first. Diff against the base branch shows only the changes in this PR.

---

In the V2 dashboard panel editor, the rules for **what works with what** were scattered across four places and two "lands":

1. `supportedSignals` — declared per kind (V2).
2. `PANEL_TYPE_TO_QUERY_TYPES` — a matrix living in **V1**'s `container/NewWidget/utils.ts
3. Query-builder field visibility — hard-coded `filterConfigs` objects.
4. The type switcher's "is this panel type allowed" logic — keyed only on the datasource.

There was **no single function** that could deterministically answer *"is this (panel kind × query type × signal) combination valid, and which query-builder fields should this kind hide?"* — and that gap produced a real bug: the type switcher only looked at the datasource, so under ClickHouse/PromQL (where there is no datasource) it left **every** type enabled, including ones that can only be authored in the Query Builder (e.g. **List**).

This PR introduces **one deterministic capability guard**. Each panel kind declares its capabilities once, next to its existing `supportedSignals`; a pure module reads the panel registry and answers every "what works with what" question. All editor consumers route through it.

**Why this approach:** it's declarative and self-extending — adding a new panel kind is one registry entry whose required fields the type system forces you to fill, and every gate (tabs, field visibility, type-switch coercion, type-switcher disabling) picks it up automatically. It also re-homes the last piece of V1 compatibility logic V2 was importing.

#### How it works

**Declaration (per kind, in `kinds/<Kind>/definition.ts`):**
```ts
supportedSignals: [metrics, logs, traces],      // existing
supportedQueryTypes: [QUERY_BUILDER, CLICKHOUSE, PROM],   // new, required
queryBuilderFields: { default: { … }, traces: { … } },   // new, optional
```
`supportedQueryTypes` is **required**, so the registry's mapped type (`{ [K in PanelKind]: PanelDefinition<K> }`) fails compilation until every kind declares it.

**The guard (`Panels/capabilities.ts`, pure):**
- `getSupportedSignals` / `isSignalSupported`
- `getSupportedQueryTypes` / `isQueryTypeSupported`
- `isPanelCombinationValid({ kind, queryType, signal })` — the master triad check
- `resolveQueryType(kind, preferred)` — coercion when switching kinds
- `getHiddenQueryBuilderFields(kind, signal)` — resolves the per-kind/per-signal field rule

**Consumers routed through it:**
- **Query-type tabs** + **query-builder field visibility** in `PanelEditorQueryBuilder` (now keyed on `PanelKind`).
- **Coercion on panel-type switch** in `usePanelTypeSwitch` (e.g. PromQL → List collapses to Query Builder).
- **Type-switcher disabling** in `PanelTypeSwitcher`: a kind is disabled when the **active query type** *or* datasource is incompatible — with a **tooltip explaining why**. The active query type is read from the query-builder provider (`currentQuery.queryType`), so a brand-new panel that hasn't staged a query yet still gates correctly.

#### Issues closed by this PR
Closes https://github.com/SigNoz/pulse-pod/issues/54
Closes https://github.com/SigNoz/pulse-pod/issues/13

---

### ✅ Change Type
_Select all that apply_

- [x] ✨ Feature
- [x] 🐛 Bug fix
- [x] ♻️ Refactor
- [ ] 🛠️ Infra / Tooling
- [ ] 🧪 Test-only

---

#### Root Cause
The visualization type switcher disabled a panel type purely on signal support (`!!signal && !supportedSignals.includes(signal)`). For ClickHouse/PromQL queries there is no datasource, so `signal` is `undefined`, the guard short-circuited, and **no** type was disabled — leaving Query-Builder-only kinds (List) and ClickHouse/PromQL-incompatible kinds (Table/Pie under PromQL) wrongly selectable.

#### Fix Strategy
Disable through `isPanelCombinationValid({ kind, queryType, signal })`, which also checks the **query type**. The live query type is sourced from the query-builder provider (not the panel spec, which lags until a query is staged) so even an empty new panel gates correctly.

---

- **Tests added/updated:**
  - `Panels/__tests__/capabilities.test.ts` — table-driven over all 7 kinds; pins query-type/signal support against the legacy matrix and the List field rules so the migration can't silently drift.
  - `PanelEditorQueryBuilder.test.tsx` — tabs + `filterConfigs` derive from the guard per kind/signal.
  - `PanelTypeSwitcher.test.tsx` — disabling under PromQL/ClickHouse and per datasource; `utils.test.ts` — the disabled-reason messages + query-type-over-signal precedence.
  - `usePanelTypeSwitch.test.ts` / `VisualizationSection.test.tsx` — updated for the guard.
- **Manual verification:** new Time Series panel → ClickHouse/PromQL tab → List/Table/Pie disabled with tooltips; metrics Query-Builder query → List disabled ("doesn't support metrics data"); switching into List from PromQL coerces the tab to Query Builder. Regression-checked the logs/traces explorers (shared `QueryBuilderV2` untouched).
- **Edge cases covered:** empty/not-yet-staged new panel; PromQL/ClickHouse (no datasource); both query-type and signal incompatible simultaneously.
- Full affected suite: **309 tests / 51 suites green**; `tsgo`, `oxlint`, `pnpm build` clean.

---

### ⚠️ Risk & Impact Assessment

- **Blast radius:** V2 dashboard panel editor only. The shared `QueryBuilderV2` and `MetricsAggregateSection` are intentionally **not** modified, so V1 widgets / explorers / alerts are unaffected. `ConfigSelect` gains an opt-in `tooltip` prop (no behavior change for existing callers).
- **Potential regressions:** the visualization type switcher's disabled set, the query-type tabs shown per kind, and List's query-builder field hiding. All covered by tests; behavior parity with the legacy matrix is asserted.
- **Rollback plan:** revert this PR; the base branch is unaffected.

---

### 📝 Changelog

| Field | Value |
|------|-------|
| Deployment Type | Cloud / OSS / Enterprise |
| Change Type | Feature / Bug Fix |
| Description | V2 dashboards: the panel editor now deterministically gates panel type × query type × signal combinations and hides inapplicable query-builder fields; incompatible visualization types are disabled with a tooltip explaining why. |

---

### 📋 Checklist
- [x] Tests added or explicitly not required
- [x] Manually tested
- [x] Breaking changes documented (none)
- [x] Backward compatibility considered (shared query builder untouched; `ConfigSelect.tooltip` is opt-in)

---

## 👀 Notes for Reviewers

- **Stacked PR:** base is `fix/dashboards-v2-panel-editor`, not `main`.
- Two commits, each independently compiling/bisectable:
  1. `feat(dashboards-v2): add a panel capability guard` — contract + per-kind declarations + guard module + tests (no consumer wired, no behavior change).
  2. `feat(dashboards-v2): gate the panel editor through the capability guard` — wires the consumers, adds query-type-aware disabling + the reason tooltip, and removes V2's last `PANEL_TYPE_TO_QUERY_TYPES` import.
- The single intentional shared-infra touch is `ConfigSelect` gaining an optional per-option `tooltip`.
- `getHiddenQueryBuilderFields` mirrors `QueryBuilderV2`'s internal List field config rather than changing it; a test asserts the two agree, keeping the guard the single source of truth without altering shared infra.

### CODE REVIEW

File: frontend/src/pages/DashboardPageV2/DashboardContainer/Panels/types/panelDefinition.ts [hunk 1]
```diff
@@ -35,7 +37,12 @@ export interface PanelDefinition<K extends PanelKind = PanelKind> {
 	displayName: string;
 	Renderer: ComponentType<PanelRendererProps<K>>;
 	sections: SectionConfig[];
+	/** Signals (datasources) this kind can visualize. */
```
Comment (AshwinBhatkal) [inline frontend/src/pages/DashboardPageV2/DashboardContainer/Panels/types/panelDefinition.ts]: no datasources 🙅🏻‍♂️

File: frontend/src/pages/DashboardPageV2/DashboardContainer/Panels/types/panelCapabilities.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import type { TelemetrytypesSignalDTO } from 'api/generated/services/sigNoz.schemas';
+import type { QueryBuilderProps } from 'container/QueryBuilder/QueryBuilder.interfaces';
+
+/**
+ * Query-builder field-visibility config a panel kind can declare, mirroring the
+ * shape `QueryBuilderV2` consumes via its `filterConfigs` prop. Derived from that
+ * prop type (the underlying `FilterConfigs` isn't exported) so the two never drift.
+ */
+export type FilterConfigsPartial = NonNullable<
```
Comment (AshwinBhatkal) [inline frontend/src/pages/DashboardPageV2/DashboardContainer/Panels/types/panelCapabilities.ts:9]: is this export required

File: frontend/src/pages/DashboardPageV2/DashboardContainer/Panels/capabilities.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+/**
+ * Query-builder field visibility for a kind + signal: the kind's `default` rule with
+ * its per-signal overrides merged over it (signal wins). `{}` when the kind declares
+ * nothing, i.e. the builder shows every field.
+ */
+export function getHiddenQueryBuilderFields(
+	kind: PanelKind,
+	signal?: TelemetrytypesSignalDTO,
```
Comment (AshwinBhatkal) [inline frontend/src/pages/DashboardPageV2/DashboardContainer/Panels/capabilities.ts]: This will always be there. Backtrack and see how to make it required

File: frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/ConfigPane/ConfigPane.tsx [hunk 4]
```diff
@@ -19,6 +20,12 @@ interface ConfigPaneProps {
 	onChangeSpec: (next: DashboardtypesPanelSpecDTO) => void;
 	/** Switch the panel to another visualization kind. */
 	onChangePanelKind: (kind: PanelKind) => void;
+	/**
+	 * Active query type from the query-builder provider (the selected tab). Drives which
+	 * panel types the visualization switcher disables — read from the provider, not the
+	 * spec, because a new panel's spec has no query until staged.
+	 */
+	queryType?: EQueryType;
```
Comment (AshwinBhatkal) [inline frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/ConfigPane/ConfigPane.tsx]: make this required everywhere?

File: frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/ConfigPane/sectionRegistry.tsx [hunk 5]
```diff
@@ -161,6 +161,9 @@ export interface ErasedSectionDescriptor {
 		// type switcher.
 		panelKind?: unknown;
 		onChangePanelKind?: unknown;
+		// Active query type; read by the visualization section's type switcher to
+		// disable kinds that can't be authored in it (List is Query-Builder-only).
+		queryType?: unknown;
```
Comment (AshwinBhatkal) [inline frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/ConfigPane/sectionRegistry.tsx]: actual type?

File: frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/PanelEditorQueryBuilder/PanelEditorQueryBuilder.tsx [hunk 6]
```diff
@@ -8,23 +8,35 @@ import { Color } from '@signozhq/design-tokens';
...
+	type PanelKind,
+} from '../../Panels/types/panelKind';
+
 import styles from './PanelEditorQueryBuilder.module.scss';
 
 interface PanelEditorQueryBuilderProps {
-	panelType: PANEL_TYPES;
+	/** The edited panel's visualization kind — drives supported query types + field visibility via the capabilities guard. */
+	panelKind: PanelKind;
+	/** The panel's current datasource; selects per-signal query-builder field rules. */
```
Comment (AshwinBhatkal) [inline frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/PanelEditorQueryBuilder/PanelEditorQueryBuilder.tsx]: no datasource. Also make it required

File: frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/PanelEditorQueryBuilder/PanelEditorQueryBuilder.tsx [hunk 7]
```diff
@@ -74,13 +89,16 @@ function PanelEditorQueryBuilder({
 		[onStageRunQuery],
 	);
 
+	// Query-builder field visibility for this kind + signal (e.g. List hides step
```
Comment (AshwinBhatkal) [inline frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/PanelEditorQueryBuilder/PanelEditorQueryBuilder.tsx]: Did not get this fully

File: frontend/src/pages/DashboardPageV2/DashboardContainer/Panels/types/panelDefinition.ts (hunk 1 above)
Comment (ahrefabhi) [reply frontend/src/pages/DashboardPageV2/DashboardContainer/Panels/types/panelDefinition.ts]: Done

File: frontend/src/pages/DashboardPageV2/DashboardContainer/Panels/capabilities.ts (hunk 3 above)
Comment (ahrefabhi) [reply frontend/src/pages/DashboardPageV2/DashboardContainer/Panels/capabilities.ts]: Done

File: frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/ConfigPane/ConfigPane.tsx (hunk 4 above)
Comment (ahrefabhi) [reply frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/ConfigPane/ConfigPane.tsx]: Done

File: frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/PanelEditorQueryBuilder/PanelEditorQueryBuilder.tsx (hunk 6 above)
Comment (ahrefabhi) [reply frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/PanelEditorQueryBuilder/PanelEditorQueryBuilder.tsx]: Done

File: frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/PanelEditorQueryBuilder/PanelEditorQueryBuilder.tsx (hunk 7 above)
Comment (ahrefabhi) [reply frontend/src/pages/DashboardPageV2/DashboardContainer/PanelEditor/PanelEditorQueryBuilder/PanelEditorQueryBuilder.tsx]: updated the comment, QB follows a config list to show/hide/disable fields this is just, this is just forwarding that config to QB

**Review by AshwinBhatkal — APPROVED** [review]
